### PR TITLE
[spanify] Use safe buffers pt.3

### PR DESCRIPTION
**Pull request by cdesouza-chromium**
This change avoids certain uses of `std::next`/`prev`, or in certain cases simplifies their use, to make it more obvious what certain ranged operations are actually doing.

Most changes here are standard. However, with `ReorderMatch` in particular, a rewrite was done, as `ReorderMatch` was to obscure, and it was doing two distinct things at the same time.

Resolves https://github.com/brave/brave-browser/issues/42489

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Review by tmancey — APPROVED** [review]

**Review by simonhong — APPROVED** [review]
++ verified autocomplete changes doesn't affect current behavior.

File: chromium_src/components/omnibox/browser/autocomplete_result.h [hunk 1]
```diff
@@ -6,9 +6,18 @@
...
+#define SortAndCull                                           \
+  Unused();                                                   \
+  void RemoveAllMatchesNotOfType(AutocompleteProvider::Type); \
+  template <typename UnaryPredicate>                          \
+  void MoveMatchToBeLast(UnaryPredicate predicate) {          \
+    if (auto it = base::ranges::find_if(matches_, predicate); \
+        it != matches_.end()) {                               \
+      std::rotate(it, std::next(it), matches_.end());         \
+    }                                                         \
+  }                                                           \
```
Comment (goodov) [inline chromium_src/components/omnibox/browser/autocomplete_result.h:20]: pls move the impl to `.cc`

File: chromium_src/components/omnibox/browser/autocomplete_result.cc [hunk 2]
```diff
@@ -18,27 +19,14 @@
...
-  } else {
-    std::rotate(it, std::next(it), std::next(pos));
-  }
-}
-
-void AutocompleteResult::RemoveMatch(const ACMatches::iterator& it) {
-  matches_.erase(it);
+void AutocompleteResult::RemoveAllMatchesNotOfType(
+    AutocompleteProvider::Type type) {
+  matches_.erase(
```
Comment (goodov) [inline chromium_src/components/omnibox/browser/autocomplete_result.cc:24]: `matches_.erase_if`?

File: chromium_src/components/omnibox/browser/autocomplete_result.h (hunk 1 above)
Comment (cdesouza-chromium) [reply chromium_src/components/omnibox/browser/autocomplete_result.h:20]: Are you sure? This is a templated function.

File: chromium_src/components/omnibox/browser/autocomplete_result.cc (hunk 2 above)
Comment (cdesouza-chromium) [reply chromium_src/components/omnibox/browser/autocomplete_result.cc]: Done.

File: chromium_src/components/omnibox/browser/autocomplete_result.h (hunk 1 above)
Comment (goodov) [reply chromium_src/components/omnibox/browser/autocomplete_result.h:20]: > Are you sure? This is a templated function.

ah, missed that

**Review by goodov — APPROVED** [review]

**Review by SergeyZhukovsky — APPROVED** [review]
wallet ++

**Pull request merged**